bundle/edu.gemini.horizons.server/src/main/scala/edu/gemini/horizons/server/backend/HS2.scala

```diff
@@ -0,0 +1,299 @@
+  // columns found). Returns a list of values or an error message.
+  private def parseMany[A](header: String, tail: List[String], headerPattern: Regex)(f: Offsets => Option[String => A] ): String \/ List[A] =
+    (headerPattern.findFirstMatchIn(header) \/> ("Header pattern not found: " + headerPattern)) *>
+    (tail.dropWhile(s => !s.trim.startsWith("---")) match {
+      case Nil                => Nil.right // no column headers means no results!
+      case colHeaders :: rows =>           // we have a header row with data rows following
+        val offsets = "-+".r.findAllMatchIn(colHeaders).map(m => (m.start, m.end)).toList
+        try {
+          f(offsets).map(g => rows.takeWhile(_.trim.nonEmpty).map(g)) \/> "Not enough columns."
+        } catch {
+          case    nfe: NumberFormatException           =>  ("Number format exception: " + nfe.getMessage).left
+          case sioobe: StringIndexOutOfBoundsException =>   "Column value(s) not found.".left
+        }
+    })
+
+  // Split into header/tail, parse
+  private def parseHeader[A](lines: List[String])(f: (String, List[String]) => String \/ List[A]): String \/ List[A] =
+    lines match {
+      case _ :: h :: t => f(h, t)
+      case _           => "Fewer than 2 lines!".left
+    }
+
+  // Parse the result of the given search
+  private def parseResponse[A](s: Search[A], lines: List[String]): \/[String, List[Row[A]]] =
+    parseHeader[Row[A]](lines) { case (header, tail) => 
+      s match {
+
+        case Search.Comet(_) => 
+
+          // Common case is that we have many results, or none.
+          lazy val case0 =
+            parseMany[Row[HorizonsDesignation.Comet]](header, tail, """  +Small-body Search Results  """.r) { os =>
+              (os.lift(2) |@| os.lift(3)).tupled.map {
+                case ((ods, ode), (ons, one)) => { row =>
+                  val desig = row.substring(ods, ode).trim
+                  val name  = row.substring(ons, one).trim
+                  Row(HorizonsDesignation.Comet(desig), name)
+                }
+              }
+            }
+
+          // Single result with form: JPL/HORIZONS      Hubble (C/1937 P1)     2015-Dec-31 11:40:21
+          lazy val case1 =
+            """  +([^(]+)\s+\((.+?)\)  """.r.findFirstMatchIn(header).map { m =>
+              List(Row(HorizonsDesignation.Comet(m.group(2)), m.group(1)))
+            } \/> "Could not match 'Hubble (C/1937 P1)' header pattern."
+
+          // Single result with form: JPL/HORIZONS         1P/Halley           2015-Dec-31 11:40:21
+          lazy val case2 =
+            """  +([^/]+)/(.+?)  """.r.findFirstMatchIn(header).map { m =>
+              List(Row(HorizonsDesignation.Comet(m.group(1)), m.group(2)))
+            } \/> "Could not match '1P/Halley' header pattern."
+
+          // First one that works!
+          case0 orElse
+          case1 orElse
+          case2 orElse "Could not parse the header line.".left
```

Comment:
Is there any benefit from clarifying "Could not parse the header line as a comet."?